source/joker_effects.c

```diff
@@ -5,180 +5,206 @@
     JokerEffect effect = {0};
-    if (scored_card == NULL) effect.mult = 4;
+
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
+
+    effect.mult = 4;
     return effect;
 }
 
-static JokerEffect sinful_joker_effect(Card *scored_card, u8 sinful_suit) {
+static JokerEffect sinful_joker_effect(Card *scored_card, u8 sinful_suit, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL && scored_card->suit == sinful_suit)
+
+    SCORE_ON_CARD_SCORED_ONLY(scored_card, scored_when, effect)
+
+    if (scored_card->suit == sinful_suit) {
         effect.mult = 3;
+    }
+
     return effect;
 }
 
-static JokerEffect greedy_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, DIAMONDS);
+static JokerEffect greedy_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, DIAMONDS, scored_when);
 }
 
-static JokerEffect lusty_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, HEARTS);
+static JokerEffect lusty_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, HEARTS, scored_when);
 }
 
-static JokerEffect wrathful_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, SPADES);
+static JokerEffect wrathful_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, SPADES, scored_when);
 }
 
-static JokerEffect gluttonous_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, CLUBS);
+static JokerEffect gluttonous_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, CLUBS, scored_when);
 }
 
-static JokerEffect jolly_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect jolly_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
 
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
+
+    // this Joker scores in an independant manner
     // This is really inefficient but the only way at the moment to check for whole-hand conditions
     u8 suits[NUM_SUITS];
     u8 ranks[NUM_RANKS];
     get_played_distribution(ranks, suits);
 
     if (hand_contains_n_of_a_kind(ranks) >= 2)
         effect.mult = 8;
+
     return effect;
 }
 
-static JokerEffect zany_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect zany_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
 
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
```

Comment:
```suggestion
    SCORE_ON_EVENT_ONLY(JOKER_CALLBACK_INDEPENDANT, scored_when, effect)
```

Batch - Consolidate joker event restriction macros